### PR TITLE
k8sjob add inspect interface for task

### PR DESCRIPTION
**Pull request by chengjoey**
#### What type of this PR

Add one of the following kinds:
/kind feature

#### What this PR does / why we need it:
report k8sjob inspect if task failed

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda-org.erda.cloud/erda/workBench/projects/387/issues/all?id=70728&issueFilter__urlQuery=eyJzdGF0ZUJlbG9uZ3MiOlsiT1BFTiIsIldPUktJTkciLCJXT05URklYIiwiUkVPUEVOIl0sImFzc2lnbmVlSURzIjpbIjEwMDEyMDUiXX0%3D&issueTable__urlQuery=eyJwYWdlTm8iOjF9&issueViewGroup__urlQuery=eyJ2YWx1ZSI6ImthbmJhbiIsImNoaWxkcmVuVmFsdWUiOnsia2FuYmFuIjoiZGVhZGxpbmUifX0%3D&iterationID=431&type=TASK)

### CODE REVIEW

File: apistructs/k8s.go [hunk 1]
```diff
@@ -13,6 +13,16 @@
...
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
 const (
 	AliyunRegistry = "aliyun-registry"
 )
+
+type TaskInspect struct {
+	Object runtime.Object    `json:"object"`
```
Comment (Effet) [inline apistructs/k8s.go:26]: type comment

File: modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/plugins/k8sflink/flink.go [hunk 2]
```diff
@@ -200,6 +201,10 @@ func (k *K8sFlink) BatchDelete(ctx context.Context, tasks []*spec.PipelineTask)
 	return nil, nil
 }
 
+func (k *K8sFlink) Inspect(ctx context.Context, task *spec.PipelineTask) (apistructs.TaskInspect, error) {
```
Comment (Effet) [inline modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/plugins/k8sflink/flink.go:204]: func comment

**Review by Effet — CHANGES_REQUESTED** [review]

File: modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/plugins/k8sflink/flink.go [hunk 3]
```diff
@@ -200,6 +201,10 @@ func (k *K8sFlink) BatchDelete(ctx context.Context, tasks []*spec.PipelineTask)
...
+	return apistructs.TaskInspect{}, errors.New("k8sflink don`t support inspect")
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/plugins/k8sflink/flink.go:205]: doesn't

File: modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/plugins/k8sjob/k8sjob.go [hunk 4]
```diff
@@ -305,6 +305,21 @@ func (k *K8sJob) BatchDelete(ctx context.Context, tasks []*spec.PipelineTask) (d
 	return nil, nil
 }
 
+func (k *K8sJob) Inspect(ctx context.Context, task *spec.PipelineTask) (apistructs.TaskInspect, error) {
+	jobName := strutil.Concat(task.Extra.Namespace, ".", task_uuid.MakeJobID(task))
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/plugins/k8sjob/k8sjob.go:309]: abstract a method for jobName.

File: modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/plugins/k8sflink/flink.go (hunk 3 above)
Comment (sfwn) [reply modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/plugins/k8sflink/flink.go:205]: `Inspect` will be invoked internally only, so no error needed?

File: modules/pipeline/pipengine/actionexecutor/plugins/scheduler/logic/events.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+		refUID = &stringRefUID
+	}
+
+	e := client.Events(ref.Namespace)
+	fieldSelector := e.GetFieldSelector(&ref.Name, &ref.Namespace, refKind, refUID)
+	initialOpts := metav1.ListOptions{FieldSelector: fieldSelector.String(), Limit: DefaultEventsLimit}
+	eventList := &corev1.EventList{}
+	err = FollowContinue(&initialOpts,
+		func(options metav1.ListOptions) (runtime.Object, error) {
+			newEvents, err := e.List(context.TODO(), options)
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/scheduler/logic/events.go:57]: context.backend

File: modules/pipeline/pipengine/actionexecutor/plugins/scheduler/scheduler.go [hunk 6]
```diff
@@ -359,6 +359,19 @@ func (s *Sched) Status(ctx context.Context, action *spec.PipelineTask) (desc api
...
+		taskExecutor   tasktypes.TaskExecutor
+		shouldDispatch bool
+		err            error
+	)
+	shouldDispatch, taskExecutor, err = s.GetTaskExecutor(action.Type, action.Extra.ClusterName, action)
+	if err != nil {
+		return nil, err
+	}
+	if !shouldDispatch {
+		logrus.Infof("task executor %s execute cancel", taskExecutor.Name())
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/scheduler/scheduler.go:372]: It doesn't matter to `cancel` ?

File: modules/pipeline/pipengine/actionexecutor/plugins/scheduler/logic/events.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+)
+
+const (
+	DefaultEventsLimit = 5
+)
+
+var metadataAccessor = meta.NewAccessor()
+
+// SearchEvents finds events about the specified object.
+func SearchEvents(client corev1client.EventsGetter, objOrRef runtime.Object) (*corev1.EventList, error) {
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/scheduler/logic/events.go:35]: how long will the method cost?

File: apistructs/pipeline_report.go [hunk 8]
```diff
@@ -50,6 +50,7 @@ var (
 	PipelineReportTypeBasic   PipelineReportType = "basic"
 	PipelineReportTypeAPITest PipelineReportType = "api-test"
 	PipelineReportTypeEvent   PipelineReportType = "event"
+	PipelineReportTypeInspect PipelineReportType = "inspect"
```
Comment (sfwn) [inline apistructs/pipeline_report.go:53]: task-inspect

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/plugins/k8sflink/flink.go (hunk 3 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/plugins/k8sflink/flink.go:205]: return error means 'taskop wait' doesn`t need to report task inspect, becase k8sflink and k8sspark don`t support inspect

File: modules/pipeline/pipengine/actionexecutor/plugins/scheduler/logic/events.go (hunk 7 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/plugins/scheduler/logic/events.go:35]: it`s similar to kubectl describe, handles the continue parameter returned by the API server when using list options

**Review by sfwn — APPROVED** [review]

**Pull request merged**